## 1.Importing Libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [2]:
pd.options.display.max_columns=None

## 2.Reading Data

In [3]:
flights = pd.read_csv('flight_price.csv')

In [4]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


- The dataset contains 10,683 rows and 11 features
- columns `Route` and `Total_stops` have missing value each
- The data types of some features isn't appropriate

## 3.Preliminary Analysis
### 3.1 check data types

In [5]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [6]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [7]:
value = flights.Date_of_Journey.iloc[4]
value

'01/03/2019'

In [8]:
value = flights.Dep_Time.iloc[4]
value

'16:50'

In [9]:
value = flights.Arrival_Time.iloc[4]
value

'21:35'

### 3.2 check for duplicates

In [10]:
flights.duplicated().sum()

220

In [11]:
# default value of keep is first.

(
    flights
    .loc[flights.duplicated(keep=False)]
    .sort_values(['Airline','Date_of_Journey','Source','Destination','Dep_Time','Duration','Total_Stops'])
    
)

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
1535   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2870    05:45         08:35   2h 50m    non-stop   
1535    20:30         23:20   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2870  No check-in baggage included   4273  
1535  No check-in baggage included   3873  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

### 3.3 observations

- The data type of `Date_of_Journey`,`Dep_Time`,`Arrival_Time` should be changed to datetime.
- The type of `Duration` & `Total_Stops` is mixed. it should be numeric type.
- There're 220 duplicates. These should be removed.

## 4.Detailed Analysis

### Airline

In [12]:
flights.Airline

0             IndiGo
1          Air India
2        Jet Airways
3             IndiGo
4             IndiGo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [13]:
flights.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [14]:
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

In [15]:
(
    flights
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
    .unique()
)

array(['Indigo', 'Air India', 'Jet Airways', 'Spicejet',
       'Multiple Carriers', 'Goair', 'Vistara', 'Air Asia', 'Trujet'],
      dtype=object)

- some of the entries have inconsistent/inaccurate values

### Date_of_Journey

In [16]:
flights.Date_of_Journey

0        24/03/2019
1         1/05/2019
2         9/06/2019
3        12/05/2019
4        01/03/2019
            ...    
10678     9/04/2019
10679    27/04/2019
10680    27/04/2019
10681    01/03/2019
10682     9/05/2019
Name: Date_of_Journey, Length: 10683, dtype: object

In [17]:
pd.to_datetime(flights.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

### Source

In [18]:
flights.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### Destination

In [19]:
flights.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

### Route
- `Route` this feature provide the same info as source to destination & total stops. so,drop it.

### Dep_Time

In [20]:
flights.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [21]:
# check correct format
# identify any feature has value apart from some other values.
# flights.Dep_Time = lambda ser

(
 flights
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

In [22]:
pd.to_datetime(flights.Dep_Time,format='%H:%M').dt.time

0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

### Arrival_Time

In [23]:
flights.Arrival_Time

0        01:10 22 Mar
1               13:15
2        04:25 10 Jun
3               23:30
4               21:35
             ...     
10678           22:25
10679           23:20
10680           11:20
10681           14:10
10682           19:15
Name: Arrival_Time, Length: 10683, dtype: object

In [24]:
(
    flights
    .Arrival_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
    .str.split(' ',n=1)
    .str.get(1)
    .unique()
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

### Duration

In [25]:
flights.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [26]:
(
    flights
    .Duration
    .loc[lambda ser : ~ser.str.contains('m')] 
    .unique()
)

array(['19h', '23h', '22h', '12h', '3h', '5h', '10h', '18h', '24h', '15h',
       '16h', '8h', '14h', '20h', '13h', '11h', '9h', '27h', '26h', '4h',
       '7h', '30h', '21h', '28h', '47h', '6h', '25h', '38h', '34h'],
      dtype=object)

In [27]:
(
    flights
    .Duration
    .loc[lambda ser : ~ser.str.contains('h')] 
    .unique()
)

array(['5m'], dtype=object)

In [28]:
(
    flights
    .Duration
    .loc[lambda ser : ~ser.str.contains('h')] 
)

6474    5m
Name: Duration, dtype: object

In [29]:
flights.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route  \
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD   

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- The observation indexed 6474 has duration of 5 min. this is clearly wrong entry. will delete this operation.

In [30]:
(
    flights
    .Duration
    .drop(index=6474)
    .str.split(' ',expand=True)
    .set_axis(['hour','minute'],axis=1)
    .assign(
        hour=lambda df_ : (
            df_
            .hour
            .str.replace('h','')
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_: (
            df_
            .minute
            .str.replace('m','')
            .fillna('0')
            .astype(int)
        )
    )
    .sum(axis=1)
    #.dtypes
    #.isna().sum()
    #.dtypes

)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [31]:
(
    flights
    .Duration
    .drop(index=6474)
    .str.split(' ',expand=True)
    .set_axis(['hour','minute'],axis=1)
    .assign(
        hour=lambda df_ : (
            df_
            .hour
            .str.replace('h','')
            .astype(int)
            .mul(60)
        ),
        minute=lambda df_: (
            df_
            .minute
            .str.replace('m','')
            .fillna('0')
            .astype(int)
        )
    )
    .sum(axis=1)
    .rename('duration_min')
    .to_frame()
    .join(flights.Duration.drop(index=6474))
)

duration_min Duration
0               170   2h 50m
1               445   7h 25m
2              1140      19h
3               325   5h 25m
4               285   4h 45m
...             ...      ...
10678           150   2h 30m
10679           155   2h 35m
10680           180       3h
10681           160   2h 40m
10682           500   8h 20m

[10682 rows x 2 columns]

### Total_Stops

In [32]:
flights.Total_Stops.unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [33]:
(
    flights
    .Total_Stops
    .replace('non-stop','0')
    .str.replace(' stops?','',regex=True)    # ' stop' or ' stops'
    .pipe(lambda ser: pd.to_numeric(ser))
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

### Additional_Info

In [34]:
flights.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

In [35]:
(
    flights
    .Additional_Info
    .str.title()
    .unique()
)

array(['No Info', 'In-Flight Meal Not Included',
       'No Check-In Baggage Included', '1 Short Layover',
       '1 Long Layover', 'Change Airports', 'Business Class',
       'Red-Eye Flight', '2 Long Layover'], dtype=object)

## 5.Cleaning Operations
- assign func used to update and create column

In [36]:
# select_dtypes(include='O') this select object datatype columns
flights.select_dtypes(include="O").columns

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info'],
      dtype='object')

In [37]:
def convert_to_min(ser):
    return (
        ser
        .str.split(' ',expand=True)
        .set_axis(['hour','minute'],axis=1)
        .assign(
                hour=lambda df_ : (
                        df_
                        .hour
                        .str.replace('h','')
                        .astype(int)
                        .mul(60)
                ),
                minute=lambda df_: (
                        df_
                        .minute
                        .str.replace('m','')
                        .fillna('0')
                        .astype(int)
                )
        )
        .sum(axis=1)                        
)


In [38]:
def clean_data(df):
    return (
            df
            .drop(index=[6474])
            .drop_duplicates()
            .assign(**{                                                          # ** dict unpack
                    col : df[col].str.strip()                                    # dict comprehension
                    for col in df.select_dtypes(include='O').columns
            })
            .rename(columns=str.lower)
            .assign(
                    airline = lambda df_: (
                            df_
                            .airline
                            .str.replace(" Premium economy","")
                            .str.replace(" Business","")
                            .str.title()                
                    ),
                    date_of_journey=lambda df_: ( 
                            pd.to_datetime(df_.date_of_journey,dayfirst=True)            
                    ),
                    dep_time=lambda df_ : (
                            pd.to_datetime(df_.dep_time,format='%H:%M').dt.time
                    ),
                    arrival_time=lambda df_ : (
                            pd.to_datetime(df_.arrival_time).dt.time
                    ),
                    duration=lambda df_: (
                            df_.duration.pipe(convert_to_min)
                    ),
                    total_stops=lambda df_: (
                            df_
                            .total_stops
                            .replace('non-stop','0')
                            .str.replace(' stops?','',regex=True)   
                            .pipe(lambda ser: pd.to_numeric(ser))
                    ),
                    additional_info=lambda df_: (
                            df_
                            .additional_info
                            .str.title()
                    )
            
                )
                .drop(columns='route')
            )

In [39]:
flights_cleaned = clean_data(flights)#.dtypes#.airline.unique()
flights_cleaned

C:\Users\DELL\AppData\Local\Temp\ipykernel_11680\1391166084.py:26: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(df_.arrival_time).dt.time


airline date_of_journey    source destination  dep_time  \
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00   
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## 6.Split the Data

In [40]:
flights_final=flights_cleaned.copy()

In [41]:
X=flights_final.drop('price',axis=1)
y=flights_final['price']

In [42]:
X_,X_test,y_,y_test = train_test_split(X,y,test_size=0.2,random_state=42)
X_train,X_val,y_train,y_val = train_test_split(X_,y_,test_size=0.2,random_state=42)

In [43]:
print(X_train.shape,y_train.shape)
print(X_test.shape,y_test.shape)
print(X_val.shape,y_val.shape)

(6695, 9) (6695,)
(2093, 9) (2093,)
(1674, 9) (1674,)


## 7.Export the subsets

In [44]:
train=X_train
train['price']=y_train
train.to_csv('train.csv',index=False)

In [45]:
test=X_test
test['price']=y_test
test.to_csv('test.csv',index=False)

In [46]:
validation=X_val
validation['price']=y_val
validation.to_csv('validation.csv',index=False)